source/mesh_refinement/isotherms.cc

```diff
@@ -0,0 +1,343 @@
+        std::vector<std::vector<double> > composition_values (quadrature.size(),
+                std::vector<double> (this->n_compositional_fields()));
+
+        typename MaterialModel::Interface<dim>::MaterialModelInputs in(quadrature.size(), this->n_compositional_fields());
+        typename MaterialModel::Interface<dim>::MaterialModelOutputs out(quadrature.size(), this->n_compositional_fields());
+
+        typename DoFHandler<dim>::active_cell_iterator
+        cell = this->get_dof_handler().begin_active(),
+        endc = this->get_dof_handler().end();
+        for (; cell!=endc; ++cell)
+            if (cell->is_locally_owned())
+            {
+                bool coarsen = false;
+                bool clear_refine = false;
+                bool refine = false;
+                bool clear_coarsen = false;
+                fe_values.reinit(cell);
+                fe_values[this->introspection().extractors.pressure].get_function_values (this->get_solution(),
+                        pressure_values);
+                fe_values[this->introspection().extractors.temperature].get_function_values (this->get_solution(),
+                        temperature_values);
+                for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                    fe_values[this->introspection().extractors.compositional_fields[c]].get_function_values (this->get_solution(),
+                            prelim_composition_values[c]);
+
+                cell->get_dof_indices (local_dof_indices);
+                in.position = fe_values.get_quadrature_points();
+                in.strain_rate.resize(0);// we are not reading the viscosity
+                for (unsigned int i=0; i<quadrature.size(); ++i)
+                {
+                    in.temperature[i] = temperature_values[i];
+                    in.pressure[i] = pressure_values[i];
+                    for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                        in.composition[i][c] = prelim_composition_values[c][i];
+                }
+                this->get_material_model().evaluate(in, out);
+
+                // all data has been gethered, now do something with it.
+
+                std::vector<std::vector<int> > isotherms_i(number_of_isotherms,std::vector<int>(2));
+                // isotherms_i[0] is the minimum, and isotherms_i[1] is the maximum cell level.
+
+                // convert max,max-1,max-2 etc and min, min-1,min-2 to numbers in integer
+                for ( unsigned int it1 = 0; it1 != isotherms_i.size(); ++it1 )
+                {
+                    if(isotherms[it1][0].compare(0,3,"max")==0) {
+                        if(isotherms[it1][0].compare(0,4,"max-")==0) {
+                            std::vector<std::string> tmpNumber = Utilities::split_string_list(isotherms[it1][0],'-');
+                            isotherms_i[it1][0]=maximum_refinement_level-Utilities::string_to_int(tmpNumber[1]);
+                        } else {
```

Comment:
This kind of code is inefficient: you are splitting strings and interpreting strings in the innermost loop over cells. This would be better handled (and lead to easier to read code) if you did it in `parse_parameters()` already, or if you need to refer to variables you don't know yet at that time, at the top of this function before you loop over cells.